### ISSUE TITLE
Hazlecast Internal Port Communication Issue

### ISSUE THREAD

**Issue by rbalabomm** [hazelcast/hazelcast#11459]
Hi,

We are using 4 node IMDG Cluster on our environment with tcp-ip configuration for internal communication between the nodes.

Below is the Hazelcast.xml configuration:
```xml
<network>
  <port auto-increment="false">5701</port>
  <join>
    <multicast enabled="false"/>
    <aws enabled="false"/>
    <tcp-ip enabled="true" connection-timeout-seconds="20">
      <member-list>
        <member>IP-Address:5701</member>
        <member>IP-Address:5701</member>
      </member-list>
    </tcp-ip>
  </join>
</network>
 ```
Once we start the Hazelcast on all the nodes, initially it was clustering and later in the logs we are seeing this message:

```
 [INFO ] 2017-09-26 20:46:58 [hz._hzInstance_1_dev.IO.thread-in-1] TcpIpConnection - [*.*.*.113]:5701 [dev] [3.9-EA] Connection[id=5, /*.*.*.113:5701->/*.*.*.113:37984, endpoint=null, alive=false, type=NONE] closed. Reason: Connection closed by the other side
```

We are not understanding why is it trying to connect to its own IP to different port?

**Comment by mmedenjak** [comment]
Hi @rbalabomm !

Looks like this is an outbound socket connection to a different client or member. 
- can you search for the `37984` port in the logs and see other log statements ?
- do you have an another member or client running on the same IP?
- can you check the logs of other members and clients and see at that time (`2017-09-26 20:46:58`) why the connection was closed? It should have the same format `Connection ... closed. Reason: ...` and the `5701` port in there

**Comment by rbalabomm** [comment]
Hi @mmedenjak 

Thanks for the quick reply!
  

- At Hazelcast IMDG startup, internally it's connecting to different ports(not 37984 each time) for its own IP, this issue was raised when we tried cluster with 2 node and 4 node cluster and even with 1 node we are getting this isssue(tries to connect itself)

- No, we dont have any client or member running on same IP

- Yes we checked the logs, 5701 port is always open on all the clustered nodes. This issue was raised only after clustering is done. Even after getting this warning, our Hazelcast Client from other application are able to retrieve the data successfully. 

When we enabled security in Hazelcast, this info becomes warning as shown below:

**[*.*.*.113]:5701 [dev] [3.9-EA] Established socket connection between /*.*.*.113:5701 and /*.*.*.113:37984
[INFO ] 2017-09-26 20:46:58 [hz._hzInstance_1_dev.IO.thread-in-1] TcpIpConnection - [*.*.*.113]:5701 [dev] [3.9-EA] Connection[id=5, /*.*.*.113:5701->/*.*.*.113:37984, endpoint=null, alive=false, type=NONE] closed. Reason: Connection closed by the other side
 javax.net.ssl.SSLHandShakeException : SSL HandShake Failed after 251 trails! -> NEED_UNWRAP**

**Comment by rbalabommala** [comment]
Hi @mmedenjak 
I guess this is internal port connectivity issue within Hazelcast and  I see that this bug is added for 3.9.1 release as per https://github.com/hazelcast/hazelcast/issues/10924.  

As, this issue is not impacting any functionality of our project,
- Shall we ignore it and move forward with production ? 
- Anytime line for 3.9.1 release?

**Comment by mmedenjak** [comment]
@rbalabomm  @rbalabommala the port will be different every time. And the reason for the SSL warning is the same as the one where SSL is not enabled.
Please check the logs of the other node and post them here. Specifically, we would need the reason the connection was closed by the other member. It should be something like `Connection ... closed. Reason: ...`.

Also, you are saying that this error occurs when there is only one member. Can you post the rest of the logs here? The member is definitely trying to connect to some other node, the question then is which node it is trying to connect to. It should not connect to itself.

You may ignore the error if it is ok with you that the connection wasn't established and you can see from the other logs that the cluster is formed with as many members as you expect. Otherwise I think you should analyse the issue.

I don't think we will be addressing this in 3.9.1 yet as we don't know enough about the issue for now.

**Comment by BioLogisch** [comment]
We have the same Issue with only one Member:

It happens every 30 seconds. 

`
2017-11-16 13:57:57,529 [info] c.h.n.t.TcpIpAcceptor hz.a44360d4-ee5e-4a79-9bab-955d6088caa6.IO.thread-Acceptor - [192.168.255.81]:5801 [dev] [3.9] Accepting socket connection from /192.168.255.81:56133
2017-11-16 13:57:57,529 [info] c.h.n.t.TcpIpConnectionManager hz.a44360d4-ee5e-4a79-9bab-955d6088caa6.cached.thread-6 - [192.168.255.81]:5801 [dev] [3.9] Established socket connection between /192.168.255.81:5801 and /192.168.255.81:56133
2017-11-16 13:57:57,530 [info] c.h.n.t.TcpIpConnection hz.a44360d4-ee5e-4a79-9bab-955d6088caa6.IO.thread-in-1 - [192.168.255.81]:5801 [dev] [3.9] Connection[id=20, /192.168.255.81:5801->/192.168.255.81:56133, endpoint=null, alive=false, type=NONE] closed. Reason: Connection closed by the other side
`

**Comment by BioLogisch** [comment]
Never mind. I have found the issue. It is the consul tcp healt check.

**Comment by Holmistr** [comment]
Hi @rbalabomm , is there any update on this issue? Are you still experiencing this issue with the latest 3.9.2?

**Comment by mmedenjak** [comment]
@rbalabomm any update on the issue?

**Comment by childnode** [comment]
I can confirm, that this issue is no longer present / seen for me on 3.9.2

the introduced `MemberAddressProvider` interface allows to hook up some custom code for bindAdress and publicAdress provider (e.g. public port from discovery/etcd if hazelcast is dockered)

With that, all things work as expected now

but still, the logs seems to be a bit quirky where outbound ports reads like incoming

**Comment by Holmistr** [comment]
@childnode Thank you very much on confirming this :) I'm closing this issue now. If you encounter any new issues or have some suggestions, feel free to open a new issue. Thanks again!
